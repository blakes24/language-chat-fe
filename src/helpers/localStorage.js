export function setLocalStorage(key, val) {
  window.localStorage.setItem(key, JSON.stringify(val));
}

export function getFromLocalStorage(key, defaultValue) {
  let value;
  try {
    value = JSON.parse(
      window.localStorage.getItem(key) || JSON.stringify(defaultValue)
    );
  } catch (e) {
    console.error(e);
    value = defaultValue;
  }
  return value;
}
