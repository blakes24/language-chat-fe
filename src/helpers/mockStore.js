import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const initialState = {
  users: {
    current: testUser,
    items: [],
    token: "",
    loading: "idle",
    error: null,
  },
  rooms: { items: [], current: null, loading: "idle", error: null },
  messages: { items: [], loading: "idle", error: null },
  partners: { items: [], loading: "idle", error: null },
};

export const store = mockStore(initialState);
