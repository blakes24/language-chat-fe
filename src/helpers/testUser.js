import UserContext from "./UserContext";

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const UserProvider = ({
  children,
  user = testUser,
  setToken = () => console.log("set token"),
}) => <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>;

export { UserProvider };
