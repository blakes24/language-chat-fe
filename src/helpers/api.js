import axios from "axios";
import { BASE_URL } from "../config";
import { getFromLocalStorage } from "./localStorage";

/** API Class: methods used to connect to the API. */

let token = getFromLocalStorage("token", "");

class ChatApi {
  static token = token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${ChatApi.token}` };
    const params = method === "get" ? data : {};
    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      let message = err.response
        ? err.response.data.error.message
        : "Server error: try again later";
      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Create a new user and get auth token. */

  static async register(userData) {
    let res = await this.request(`auth/register`, userData, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get user details. */

  static async getUser(userId) {
    let res = await this.request(`users/${userId}`);
    return res.user;
  }

  /** Verifies that the social token provided is valid */

  static async verify(token) {
    let res = await this.request(`auth/validate`, token, "post");
    return res;
  }

  /** Get gets token to authenticate future requests to api */

  static async getToken(data) {
    let res = data.email
      ? await this.request(`auth/token`, data, "post")
      : await this.request(`auth/social-token`, data, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get all users, filter by speaks or learning language */

  static async getAllUsers(filters) {
    let res = await this.request(`users`, filters);
    return res.users;
  }

  /** Get a users chat rooms, filter by chat partner */

  static async getRooms({ userId, partnerId }) {
    let queryParams = { partner: partnerId };
    let res = await this.request(`users/${userId}/rooms`, queryParams);
    return res;
  }

  /** Creates a chat room for two users data = {user1, user2} */

  static async createRoom(data) {
    let res = await this.request(`rooms`, data, "post");
    return res;
  }

  /** Gets messages, filter by room */

  static async getMessages(roomId) {
    let queryParams = { room: roomId };
    let res = await this.request(`messages`, queryParams);
    return res.messages;
  }

  /** Update user info. */

  static async updateUser(id, data) {
    let res = await this.request(`users/${id}`, data, "patch");
    return res.user;
  }

  /** Update language(speaks). */

  static async updateSpeaks(id, data) {
    let res = await this.request(`languages/speaks/${id}`, data, "patch");
    return res.language;
  }

  /** Update language(learning). */

  static async updateLearning(id, data) {
    let res = await this.request(`languages/learning/${id}`, data, "patch");
    return res.language;
  }

  /** Get a users partners */

  static async getPartners(userId) {
    let res = await this.request(`users/${userId}/partners`);
    return res.partners;
  }

  /** Adds a partner */

  static async addPartner(userId, data) {
    let res = await this.request(`users/${userId}/partners`, data, "post");
    return res;
  }

  /** Deletes a partner */

  static async deletePartner({ userId, partnerId }, data) {
    let res = await this.request(
      `users/${userId}/partners/${partnerId}`,
      data,
      "delete"
    );
    return res;
  }

  /** Deletes a user */

  static async deleteUser(userId, data) {
    let res = await this.request(`users/${userId}`, data, "delete");
    return res;
  }
}

export default ChatApi;
