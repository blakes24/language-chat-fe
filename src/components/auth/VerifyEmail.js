import Container from "@material-ui/core/Container";
import { makeStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import { useParams, useHistory, Link } from "react-router-dom";
import ChatApi from "../../helpers/api";
import { setToken } from "../../store/usersSlice";
import { useDispatch, useSelector } from "react-redux";
import { setLocalStorage } from "../../helpers/localStorage";
import { useState } from "react";

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1,
    marginTop: "2rem",
    textAlign: "center",
  },
  btn: {
    marginTop: "1rem",
  },
}));

function VerifyEmail() {
  const classes = useStyles();
  const { code } = useParams();
  const dispatch = useDispatch();
  const history = useHistory();
  const user = useSelector((state) => state.users.current);
  const [failed, setFailed] = useState(false);
  const [success, setSuccess] = useState(false);

  async function verify() {
    try {
      let res = await ChatApi.verifyEmail(code);
      setLocalStorage("token", res);
      dispatch(setToken(res));
      setSuccess(true);
    } catch (err) {
      setFailed(true);
    }
  }
  async function resend() {
    try {
      await ChatApi.resendEmail(user.id);
      setFailed(false);
      history.push("/verify");
    } catch (err) {
      return <p>{err[0]}</p>;
    }
  }

  const successMsg = (
    <div className={classes.root}>
      <p>Your account has been verified.</p>
      <Link to="/">Start Chatting</Link>
    </div>
  );

  const failedMsg = (
    <div className={classes.root}>
      <p>Verification link is expired or invalid.</p>
      {user ? (
        <Button
          variant="contained"
          color="primary"
          onClick={resend}
          className={classes.btn}
        >
          Resend Link
        </Button>
      ) : (
        <p>
          <Link to="/login">Login</Link> to resend link.
        </p>
      )}
    </div>
  );

  if (failed) return failedMsg;
  if (success) return successMsg;

  return (
    <Container maxWidth="sm" className={classes.root}>
      <Typography gutterBottom variant="h5" component="h1" align="center">
        Welcome to LangChat
      </Typography>
      {code ? (
        <>
          <p>Click the button below to verify your account.</p>
          <Button
            variant="contained"
            color="primary"
            onClick={verify}
            className={classes.btn}
          >
            Verify
          </Button>
        </>
      ) : (
        <>
          <p>
            A verification link has been sent to <b>{user.email}</b>. Please
            follow the link in the email to verify your account.
          </p>
          <p>
            If you did not receive an email, click the button below to send a
            new verification link.
          </p>
          <Button
            variant="contained"
            color="primary"
            onClick={resend}
            className={classes.btn}
          >
            Resend Link
          </Button>
        </>
      )}
    </Container>
  );
}

export default VerifyEmail;
