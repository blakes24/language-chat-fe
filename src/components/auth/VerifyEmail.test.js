import { render } from "@testing-library/react";
import VerifyEmail from "./VerifyEmail";
import { MemoryRouter, Route } from "react-router";
import { store } from "../../helpers/mockStore";
import { Provider } from "react-redux";

test("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter initialEntries={["verify/abcd"]}>
        <Route path="verify/:code">
          <VerifyEmail />
        </Route>
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={["verify/abcd"]}>
        <Route path="verify/:code">
          <VerifyEmail />
        </Route>
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
