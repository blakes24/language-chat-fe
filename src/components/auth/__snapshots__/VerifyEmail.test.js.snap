// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-root-3 MuiContainer-maxWidthSm"
  >
    <h1
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom MuiTypography-alignCenter"
    >
      Welcome to LangChat
    </h1>
    <p>
      Click the button below to verify your account.
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btn-4 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Verify
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
